File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/TopInterview2Round/connect_116.java
package algo.tzashinorpu.SpecialSubject.TopInterview2Round;

import algo.tzashinorpu.SpecialSubject.Node;

public class connect_116 {
    public Node connect(Node root) {
        if (root == null) return null;
        helper(root.left, root.right);
        return root;
    }

    private void helper(Node left, Node right) {
        if (left == null) return;
        left.next = right;
        helper(left.left, left.right);
        helper(left.right, right.left);
        helper(right.left, right.right);
    }

}

File: Clion/ZeroTrac/Tree/labuladong/connect_116.cpp
//
// Created by TzashiNorpu on 3/14/2023.
//
#include "tree_r1.h"

using namespace Tree;

Node *Solution::connect(Node *root) {
  function<void(Node *, Node *)> helper = [&](Node *l, Node *r) -> void {
      if (!l) return;
      l->next = r;
      helper(l->left, l->right);
      helper(l->right, r->left);
      helper(r->left, r->right);
  };
  if (!root) return root;
  helper(root->left, root->right);
  return root;
}